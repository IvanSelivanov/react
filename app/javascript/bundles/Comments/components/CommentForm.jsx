import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { createComment, fieldChange, reply } from '../actions'

export class CommentFormP extends React.Component {
  static get propTypes(){
    return {
      photo: PropTypes.string,
      author: PropTypes.string,
      text: PropTypes.string,
    }
  }

  createComment = () => {
    this.props.createComment({
      photo_id: this.props.photo,
      parent_id: this.props.parent,
      comment: {
        author: this.props.author,
        text: this.props.text
      }
    })
  }

  updateValue = (e) => {
    this.props.fieldChange(e.target.name, e.target.value)
  }

  reply = () => {
    this.props.reply(this.props.parent)
  }

  render() {
    return (
      <div className="clearfix">
        <small
          className={this.props.reply_to===this.props.parent ? 'd-none' : 'd-block mb-4'}
          onClick={this.reply}
        >Ответить</small>
        <div className={this.props.reply_to!==this.props.parent ? 'd-none' : 'd-block card px-3 mb-3'} style={{width: '28rem'}}>
          <div className="form-group">
            <label>Имя:</label>
            <input
              type="text"
              className="form-control"
              name="comment_author"
              value={this.props.author}
              onChange={this.updateValue}
            />
            <small className="form-text text-muted">Представьтесь, пожалуйста</small>
          </div>
          <div className="form-group">
            <label>Комментарий:</label>
            <textarea className="form-control" name={'comment_text'} value={this.props.text} onChange={this.updateValue} />
            <input type="hidden" name="photo_id" value={this.props.photo} />
          </div>
          <div className="form-group">
            <button className="btn btn-primary" onClick={this.createComment}>Сохранить</button>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
//  console.log(state)
  return(
    {
      photo: state.get('photo'),
      author: state.get('comment_author'),
      text: state.get('comment_text'),
      reply_to: state.get('reply_to')
    }
  )
}

export const CommentForm = connect(mapStateToProps, { createComment, fieldChange, reply })(CommentFormP)