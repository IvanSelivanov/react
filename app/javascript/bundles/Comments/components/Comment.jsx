import React from 'react'
import { CommentForm } from './CommentForm'
import { CommentListP } from './CommentList'
import propTypes from 'prop-types'
import { List } from 'immutable'

export class Comment extends React.Component {
  static get propTypes(){
    return {
      text: propTypes.string,
      author: propTypes.string,
      id: propTypes.string,
      children: propTypes.instanceOf(List)
    }
}
  render() {
    return (
      <li className="media">
        <div className="media-body">
        <blockquote className="blockquote">
          <p>{this.props.text}</p>
          <footer className="blockquote-footer">{this.props.author}</footer>
        </blockquote>
        <CommentForm parent={this.props.id}/>
        <CommentListP comments={this.props.children}/>
        </div>
      </li>
    )
  }
}