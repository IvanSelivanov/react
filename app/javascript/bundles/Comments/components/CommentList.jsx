import React from 'react'
import { connect } from 'react-redux'
import { Comment } from './Comment'
import propTypes from 'prop-types'
import { List } from 'immutable'

export class CommentListP extends React.Component {
  static get propTypes(){
    return {
      comments: propTypes.instanceOf(List)
    }
  }

  render() {
    const { comments } = this.props
    return (
      <ul className="list-unstyled ml-5">
        { comments.map( comment => {
          return(
              <Comment
                key = {comment.get('id')}
                author = {comment.get('author')}
                text = {comment.get('text')}
                id = {comment.get('id')}
                children = {comment.get('comments')}
              />
          )
        })}
      </ul>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    comments: state.get('comments')
  }
}

export const CommentList = connect(mapStateToProps)(CommentListP)