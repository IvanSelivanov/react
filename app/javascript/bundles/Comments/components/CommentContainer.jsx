import React from 'react'
import { CommentForm } from './CommentForm'
import { CommentList } from './CommentList'

export class CommentContainer extends React.Component {
  render(){
    return (
      <div>
        <CommentForm/>
        <CommentList/>
      </div>
    )
  }
}