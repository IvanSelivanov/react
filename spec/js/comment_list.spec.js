import { CommentListP } from '../../app/javascript/bundles/Comments/components/CommentList'
import { comments } from './comments'
import React from 'react'
import { shallow } from 'enzyme'

describe('CommentListP', ()=>{
  let wrap = shallow(<CommentListP comments={comments} />)
  it('Matches shapshot', ()=>{
    console.log(wrap.debug())
    expect(wrap).toMatchSnapshot()
  })
  it('Has two top-level comments', ()=>{
    expect(wrap.children().length).toEqual(2)
  })
})