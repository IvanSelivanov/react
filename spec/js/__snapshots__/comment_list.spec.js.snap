// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentListP Matches shapshot 1`] = `
<ul
  className="list-unstyled ml-5"
>
  <Comment
    author="1"
    id="5a588dd5dacf99164b5ddead"
    key="5a588dd5dacf99164b5ddead"
    text="123"
  >
    id
    5a588ddddacf99164b5ddeae
    author
    er
    text
    fghf
    parent_type
    Comment
    parent_id
    5a588dd5dacf99164b5ddead
    comments
    id
    5a588ed1dacf99164b5ddeb3
    author
    rt
    text
    ewrtewr
    parent_type
    Comment
    parent_id
    5a588ddddacf99164b5ddeae
    comments
    id
    5a589761dacf99226ebba828
    author
    tyd
    text
    ytdk
    parent_type
    Comment
    parent_id
    5a588dd5dacf99164b5ddead
    comments
  </Comment>
  <Comment
    author="dsfgf"
    id="5a588e4ddacf99164b5ddeaf"
    key="5a588e4ddacf99164b5ddeaf"
    text="gdhjg"
  >
    id
    5a588e54dacf99164b5ddeb0
    author
    fgdjh
    text
    ghj
    parent_type
    Comment
    parent_id
    5a588e4ddacf99164b5ddeaf
    comments
  </Comment>
</ul>
`;
